skimage/feature/corner.py

```diff
@@ -956,38 +976,53 @@ def corner_peaks(image, min_distance=1, threshold_abs=None, threshold_rel=None,
            [2, 3],
            [2, 2]])
     >>> corner_peaks(response, threshold_rel=0)
-    array([[2, 2]])
+    array([[3, 3]])
 
     """
     if threshold_rel is None:
         threshold_rel = 0.1
         warn("Until version 0.16, threshold_rel was set to 0.1 by default. "
              "Starting from version 0.16, the default value is set to None. "
              "Until version 0.18, a None value corresponds to a threshold "
-             "value of 0.1. "
-             "The default behavior will match skimage.feature.peak_local_max. "
-             "To avoid this warning, set threshold_rel=0.",
-             category=FutureWarning, stacklevel=2)
-
-    peaks = peak_local_max(image, min_distance=min_distance,
-                           threshold_abs=threshold_abs,
-                           threshold_rel=threshold_rel,
-                           exclude_border=exclude_border,
-                           indices=False, num_peaks=num_peaks,
-                           footprint=footprint, labels=labels,
-                           num_peaks_per_label=num_peaks_per_label)
-    if min_distance > 0:
-        coords = np.transpose(peaks.nonzero())
-        for r, c in coords:
-            if peaks[r, c]:
-                peaks[max((r - min_distance), 0):r + min_distance + 1,
-                      max((c - min_distance), 0):c + min_distance + 1] = False
-                peaks[r, c] = True
-
-    if indices is True:
-        return np.transpose(peaks.nonzero())
-    else:
-        return peaks
+             "value of 0.1. The default behavior will match "
+             "skimage.feature.peak_local_max.", category=FutureWarning,
+             stacklevel=2)
+
+    if np.isinf(num_peaks):
+        num_peaks = None
+
+    # Get the coordinates of the detected peaks
+    coords = peak_local_max(image, min_distance=min_distance,
+                            threshold_abs=threshold_abs,
+                            threshold_rel=threshold_rel,
+                            exclude_border=exclude_border,
+                            indices=True, num_peaks=np.inf,
+                            footprint=footprint, labels=labels,
+                            num_peaks_per_label=num_peaks_per_label)
+
+    if len(coords):
+        # Use KDtree to find the peaks that are too close to each other
+        tree = spatial.cKDTree(coords)
+
+        rejected_peaks_indices = set()
+        for idx, point in enumerate(coords):
+            if idx not in rejected_peaks_indices:
+                candidates = tree.query_ball_point(point, r=min_distance,
+                                                   p=p_norm)
+                candidates.remove(idx)
+                rejected_peaks_indices.update(candidates)
+
+        # Remove the peaks that are too close to each other
+        coords = np.delete(coords, tuple(rejected_peaks_indices),
+                           axis=0)[:num_peaks]
```

Comment:
Sorry to chime in this late in the discussion. I think this may be closely related to #4165. As I understand the distance metric aspect here the output `coords` of `peak_local_max` is a coordinate array of the corner peaks which are then filtered with the help of cKDTree. If that is the case the proposed function `remove_near_objects` in #4165 should be able to replace this part. 

Performance wise I'm not sure which solution would be most appropriate here:
- `remove_near_objects` would require a temporary conversion of `coords` to a boolean mask (`indices=False`) and comes with additional overhead which is not needed for this use case.
- but its speed should scale significantly better with the number of peaks because most of the iteration is done in Cython. 

It could very well be that the break-even point is way beyond what is considered an expected number of peaks here so take this suggestion with a grain of salt until I or someone else had the chance to benchmark this. 

And in any case this shouldn't hold up this PR as it could be done afterwards without deprecations (if we manage to deal with the ordering of the peaks). The way I see it this is already an improvement over the old state. :+1:

Reply:
I also did not found time to review your PR #4165, so don't worry ;-)
You are right

Reply:
Yes, the order by which objects (in this case corner peaks) are removed is determined by the `priority` attribute. It defaults to the sample count of objects (which in this case should be 1 for every object?) but it would be trivial to pass in an array with the correct amplitude values. That would be the "corner measure response image" correct?

Reply:
Even if this cannot utilize another function, it should probably be factored out to shared utilities: it seems very useful!

Reply:
In any case something for another PR I'd say.